display_list/testing/dl_test_surface_metal.cc

```diff
@@ -54,5 +57,71 @@ std::shared_ptr<DlSurfaceInstance> DlMetalSurfaceProvider::MakeOffscreenSurface(
   return std::make_shared<DlMetalSurfaceInstance>(std::move(surface));
 }
 
+class DlMetalPixelData : public DlPixelData {
+  using MetalScreenshot = impeller::testing::MetalScreenshot;
+
+ public:
+  explicit DlMetalPixelData(std::unique_ptr<MetalScreenshot> screenshot)
+      : screenshot_(std::move(screenshot)),
+        addr_(reinterpret_cast<const uint32_t*>(screenshot_->GetBytes())),
+        ints_per_row_(screenshot_->GetBytesPerRow() / 4) {
+    FML_DCHECK(screenshot_->GetBytesPerRow() == ints_per_row_ * 4);
+  }
+  ~DlMetalPixelData() override = default;
+
+  const uint32_t* addr32(int x, int y) const override {
+    return addr_ + (y * ints_per_row_) + x;
+  }
+  size_t width() const override { return screenshot_->GetWidth(); }
+  size_t height() const override { return screenshot_->GetHeight(); }
+  void write(const std::string& path) const override {
+    screenshot_->WriteToPNG(path);
+  }
+
+ private:
+  std::unique_ptr<MetalScreenshot> screenshot_;
+  const uint32_t* addr_;
+  const uint32_t ints_per_row_;
+};
+
+sk_sp<DlPixelData> DlMetalSurfaceProvider::ImpellerSnapshot(
```

Comment:
I don't think you need to use the metal screenshotter here. provided you have access to an impeller context, you should be able to read back the texture data without worrying about timing issues either.

See https://github.com/flutter/engine/blob/main/lib/ui/painting/image_encoding_impeller.cc#L108-L129 


I'm also adding a helper method to do this but its still WIP:  https://github.com/flutter/engine/blob/main/lib/ui/painting/image_encoding_impeller.cc#L108-L129

Reply:
you'd need to use something like fml::WaitableEvent as a latch as the submission callback will be invoked later on a different thread

Reply:
Those point to the same code. When will the helper method be available? Is it worth using ScreenShotter until then as a temporary measure?

Reply:
Oops! helper is https://github.com/flutter/engine/pull/46394 . Uh, won't be done today for sure. But you can use the code from image_encoding_impeller today and then this works on all backends.

Reply:
I looked into bringing over the changes there and there were quite a few dependencies I had to resolve. Would it be possible to just put this in as is using the ScreenShotter and then shift over to the new convenience mechanism when that lands?

Even though this doesn't get run by default right now, it does identify a number of Impeller bugs related to ImageFilter bounds and ignoring ImageFilters, so it would be useful to have this tool in the tree so that we can start fixing those problems. When the failing tests go to zero we can look at how to get it running under CI as a regression test. Many of the failures are dups (if we don't do something under one rendering op then we don't do it on many of them) so the number of problems that actually need fixing is less than the number of failing cases. I looked at the output and saw 3 or 4 problems that kept triggering over and over, but that's not an exhaustive search...

Reply:
I think that is  fine. we should probably just refactor the screenshotter to have implementations for all platforms